File: trunk/OscarDotNet/component/biz/Class_biz_milestones.pas
unit Class_biz_milestones;

interface

const
  ID = '$Id$';

type
  TClass_biz_milestones = class
  private
    { Private Declarations }
  public
    constructor Create;
    procedure Sweep;
  end;

type
  milestone_type =
    (
    FIRST_MILESTONE = 1,
    SECOND_MILESTONE = 2
    );
implementation

uses
  Class_db_milestones,
  Class_biz_accounts,
  system.collections;

//
// REMINDER_CONTROL_TABLE
//
const
  MAX_NUM_REMINDERS = 6;
type
  relative_day_num_array_index_type = 0..5;
  relative_day_num_array_type = array[relative_day_num_array_index_type] of cardinal;
  reminder_control_record_type =
    RECORD
    num_reminders: cardinal;
    relative_day_num_array: relative_day_num_array_type;
    END;
  reminder_control_table_type = array[milestone_type] of reminder_control_record_type;
const
  REMINDER_CONTROL_TABLE: reminder_control_table_type =
    (
    (num_reminders:6; relative_day_num_array:(1,3,7,14,30,90)), // FIRST_MILESTONE
    (num_reminders:6; relative_day_num_array:(1,3,7,14,30,90))  // SECOND_MILESTONE
    );

constructor TClass_biz_milestones.Create;
begin
  inherited Create;
  // TODO: Add any constructor code here
end;

procedure TClass_biz_milestones.Sweep;
var
  be_handled: boolean;
  be_processed: boolean;
  biz_accounts: TClass_biz_accounts;
  deadline: datetime;
  db_milestones: TClass_db_milestones;
  i: cardinal;
//  j: cardinal;
  master_id: string;
  master_id_q: queue;
  milestone: milestone_type;
  relative_day_num: cardinal;
  today: datetime;
begin
  biz_accounts := TClass_biz_accounts.Create;
  db_milestones := TClass_db_milestones.Create;
  master_id_q := nil;
  today := datetime.Today;
  //
  for milestone := Low(milestone_type) to High(milestone_type) do begin
    db_milestones.Check(ord(milestone),be_processed,deadline);
    if not be_processed then begin
      if (today > deadline) then begin
        case milestone of
        FIRST_MILESTONE:
          BEGIN
          END;
        SECOND_MILESTONE:
          BEGIN
          // Not application-enforceable
          END;
        end;
        for i := 1 to master_id_q.Count do begin
          master_id := master_id_q.Dequeue.tostring;
          //biz_accounts.MakeDeadlineFailureNotification;
        end;
        db_milestones.MarkProcessed(ord(milestone));
      end else begin
        be_handled := FALSE;
        i := 0;
        while (not be_handled) and (i < REMINDER_CONTROL_TABLE[milestone].num_reminders) do begin
          relative_day_num := REMINDER_CONTROL_TABLE[milestone].relative_day_num_array[i];
          if today = deadline.AddDays(-(relative_day_num)).Date then begin
//            master_id_q := biz_emsof_requests.SusceptibleTo(milestone);
//            for j := 1 to master_id_q.Count do begin
//              master_id := master_id_q.Dequeue.tostring;
//              biz_accounts.Remind(milestone,relative_day_num,deadline,biz_emsof_requests.MemberIdOfMasterId(master_id));
//              be_handled := TRUE;
//            end;
          end;
          i := i + 1;
        end;
      end;
    end;
  end;
end;

end.

File: trunk/OscarDotNet/component/biz/Class_biz_milestones.cs
using Class_biz_users;
using Class_db_milestones;
using System;
using System.Collections;

namespace Class_biz_milestones
  {
  public class TClass_biz_milestones
    {

    private static class Static
      {
      private static reminder_control_record_type[] the_reminder_control_table =
        {
        new reminder_control_record_type {num_reminders = 6, relative_day_num_array = new uint[MAX_NUM_REMINDERS] {1, 3, 7, 14, 30, 90}},
        new reminder_control_record_type {num_reminders = 6, relative_day_num_array = new uint[MAX_NUM_REMINDERS] {1, 3, 7, 14, 30, 90}}
        };
      public const int MAX_NUM_REMINDERS = 6;
      public static reminder_control_record_type[] REMINDER_CONTROL_TABLE
        {
        get => the_reminder_control_table;
        set => the_reminder_control_table = value;
        }
      }

    private struct reminder_control_record_type
      {
      public uint num_reminders;
      public uint[] relative_day_num_array;
      }

    private enum milestone_type
      {
      FIRST_MILESTONE = 1,
      SECOND_MILESTONE = 2,
      }

    public TClass_biz_milestones() : base()
      {
      }

        public void Sweep()
        {
            bool be_handled;
            bool be_processed;
            TClass_biz_users biz_users;
            DateTime deadline;
            uint i;
            // j: cardinal;
            string master_id;
            Queue master_id_q;
            uint relative_day_num;
            DateTime today;
            biz_users = new TClass_biz_users();
            var db_milestones = new TClass_db_milestones();
            master_id_q = null;
            today = DateTime.Today;
            foreach (milestone_type milestone in Enum.GetValues(typeof(milestone_type)))
            {
                db_milestones.Check((uint)(milestone), out be_processed, out deadline);
                if (!be_processed)
                {
                    if ((today > deadline))
                    {
                        switch(milestone)
                        {
                            case milestone_type.FIRST_MILESTONE:
                                break;
                            case milestone_type.SECOND_MILESTONE:
                                break;
                        // Not application-enforceable
                        }
                        uint master_id_q_count = (uint)(master_id_q.Count);
                        for (i = 1; i <= master_id_q_count; i ++ )
                        {
                            master_id = master_id_q.Dequeue().ToString();
                        // biz_users.MakeDeadlineFailureNotification;
                        }
                        db_milestones.MarkProcessed((uint)(milestone));
                    }
                    else
                    {
                        be_handled = false;
                        i = 0;
                        while (!be_handled && (i < Static.REMINDER_CONTROL_TABLE[(int)milestone].num_reminders))
                        {
                            relative_day_num = Static.REMINDER_CONTROL_TABLE[(int)milestone].relative_day_num_array[i];
                            if (today == deadline.AddDays( -relative_day_num).Date)
                            {
                            // master_id_q := biz_emsof_requests.SusceptibleTo(milestone);
                            // for j := 1 to master_id_q.Count do begin
                            // master_id := master_id_q.Dequeue.ToString();
                            // biz_users.Remind(milestone,relative_day_num,deadline,biz_emsof_requests.MemberIdOfMasterId(master_id));
                            // be_handled := TRUE;
                            // end;
                            }
                            i++;
                        }
                    }
                }
            }
        }

    } // end TClass_biz_milestones

}
